src/game/editor/editor2.cpp

```diff
@@ -2750,6 +2757,265 @@ void CEditor2::RenderPopupBrushPalette()
 	RenderBrush(m_UiMousePos);
 }
 
+void CEditor2::InvokePopupFileSelect(const char *pInitialPath, bool (*pfnCallback)(const char *pFilename, void *pContext), void *pContext)
+{
+	m_UiFileSelectState.m_Selected = -1;
+	m_UiFileSelectState.m_pContext = pContext;
+	m_UiFileSelectState.m_pfnMapSelectCB = pfnCallback;
+	str_copy(m_UiFileSelectState.m_aPath, pInitialPath, sizeof(m_UiFileSelectState.m_aPath));
+	m_UiFileSelectState.m_ppListBoxEntries = 0;
+
+	m_UiFileSelectState.PopulateFileList(Storage(), IStorage::TYPE_SAVE);
+
+	m_UiCurrentPopupID = POPUP_FILE_SELECT;
+}
+
+int CEditor2::CUIFileSelect::EditorListdirCallback(const char *pName, int IsDir, int StorageType, void *pUser)
+{
+	CUIFileSelect *pState = (CUIFileSelect *)pUser;
+
+	if(!str_comp(pName, "."))
+		return 0;
+
+	CEditor2::CFileListItem Item;
+	str_copy(Item.m_aFilename, pName, sizeof(Item.m_aFilename));
+	if(IsDir)
+		str_format(Item.m_aName, sizeof(Item.m_aName), "%s/", pName);
+	else
+		str_copy(Item.m_aName, pName, sizeof(Item.m_aName));
+
+	Item.m_IsDir = IsDir != 0;
+	Item.m_IsLink = false;
+	Item.m_StorageType = StorageType;
+	pState->m_aFileList.add(Item);
+
+	return 0;
+}
+
+void CEditor2::CUIFileSelect::PopulateFileList(IStorage *pStorage, int StorageType)
+{
+	m_aFileList.clear();
+	pStorage->ListDirectory(StorageType, m_aPath, EditorListdirCallback, this);
+	GenerateListBoxEntries();
+
+	pStorage->GetCompletePath(IStorage::TYPE_SAVE, m_aPath, m_aCompletePath, sizeof(m_aCompletePath));
+}
+
+void CEditor2::CUIFileSelect::GenerateListBoxEntries()
+{
+	// An array<const char *> might help get allocations down
+	if(m_ppListBoxEntries)
+		free(m_ppListBoxEntries);
+
+	m_ppListBoxEntries = (const char **)mem_alloc(m_aFileList.size() * 3 * sizeof(*m_ppListBoxEntries), 0);
+	for(int i = 0; i < m_aFileList.size(); i++)
+	{
+		m_ppListBoxEntries[i * 3] = m_aFileList[i].m_aName;
+		m_ppListBoxEntries[i * 3 + 1] = "D1";
+		m_ppListBoxEntries[i * 3 + 2] = "D2";
```

Comment:
I like it when the type of a structure says how it is structured. Would it be better to type `m_ppListBoxEntries` as a `const char* [][3]` or a struct array? It looks temporary so I guess you already had plans for it

Reply:
Yeah this is definitely going to become a struct array, I'm just not completely certain how to handle the column count being able to change. I was thinking templates maybe. It's really not easy to get this right without reflection and introspection.